### DISCUSSION TITLE
Circular Reasoning in the Documentation

### DISCUSSION THREAD

**Discussion by charlestbell** [react-hook-form/react-hook-form#9088] [Q&A] [answered]
On the [useFieldArray page](https://react-hook-form.com/api/usefieldarray), under "update" it says "please use **'setValue'** API instead." 

If you go the[ setValues page](https://react-hook-form.com/api/useform/setvalue), it says "It's recommended to use field array's methods such as "replace" or **"update"** instead."

So if both Update, and setValue say they shouldn't be used with Field Arrays, does that mean "update" is the only sanctioned option? It's confusing.

Also, I don't want to use replace, because it makes react-native-draggable-flatlist crash.

**Comment by bluebill1049** [comment, marked as answer]
That's good feedback. I will address that today. 

- setValue: doesn't cause input from unmount and remount focus on a single field.
- update: will update a group of field array and unmount and remount them.

**Reply by bluebill1049** [reply]
updated: https://github.com/react-hook-form/documentation/commit/15ae8b2759a0b94713b3905dc59338c16994b1cc let me know if this improved on it, or anything else should i address.

**Reply by charlestbell** [reply]
Thanks for those updates. I think I get it now. 

I have a RN app that has a series of checkboxes in a useFieldArray, and a "Select all" checkbox that I want to update all of them.

Am I correct that the most stable, and fewest-render solution is to watch the field array data with a watch (in a useEffect), and iterate over the Field Array data and call setValue on each one?

**Reply by bluebill1049** [reply]
> Am I correct that the most stable, and fewest-render solution is to watch the field array data with a watch (in a useEffect), and iterate over the Field Array data and call setValue on each one?

could you do that with `onChange` (event) instead of `useEffect`?

**Reply by charlestbell** [reply]
Update: It works! I used the setValue syntax from your update and here's what I came up with:

```
  const components = watch('components');
  const selectAll = watch('allChecked');

  useEffect(() => {
    components.forEach((component, index) => {
      setValue(`components[${index}].isChecked`, selectAll);
    });
  }, [selectAll]);
```

It updates all the checkboxes when I hit the "Select All" checkbox, and no crashes!

I know you said in your [useWatch Video](https://youtu.be/3qLd69WMqKk?t=569) that using useEffect in this way is an antipattern, but I can't figure out how to do it another way. Any advice?

**Reply by bluebill1049** [reply]
> It updates all the checkboxes when I hit the "Select All" checkbox,

```tsx
<Button onClick={() => {
    components.forEach((component, index) => {
      setValue(`components[${index}].isChecked`, on);
    });
}}>Check All</Button>
```

it's not an anti pattern, person,ally i prefer to use from event instead of `useEffect`, it's usecase is case by case as well, see if you do what I have above, otherwise, it's fine to use what you did there too. 👍

**Reply by charlestbell** [reply]
You're an awesome maintainer! 

I made a helper function that I passed into the 'master' checkbox, to control all the other checkboxes:

```
const components = watch('components');

const handleToggleAll = useCallback(selectAll => {
    components.forEach((component, index) => {
      setValue(`components[${index}].isChecked`, selectAll);
    });
  });
```

Works like a dream.

Donation made

**Reply by bluebill1049** [reply]
amazing! remove `useCallback`, if you don't need it too :)

**Reply by bluebill1049** [reply]
oh and thanks very much for the donation! ❤️
